### PR TITLE
irjit: Optimize out more temps and lwl/lwr operations

### PR DESCRIPTION
**Pull request by unknownbrackets**
This moves to using a Load32Left and related op, instead of generating the masking right then and there.  It still generates the masking before execution, just as an optimization pass.

The pass will be skipped when using a breakpoint, but then we can just leave the backend to interp fallback.  This way we never have to reimplement these in a backend, since they're annoying.

On the way, I optimized out two cases, which happened with lwr but also happened generally.  Most specifically, this:

```
or t0, a0, a1
beq t0, v0, foo
lui t0, 0x1234
```

Would previously turn into:
```
Or t0, a0, a1
Mov lhs, t0
Mov rhs, v0
SetConst t0, 0x12340000
ExitIfEq lhs, rhs, foo
```

Now:
```
Or lhs, a0, a1
SetConst t0, 0x12340000
ExitIfEq lhs, v0, foo
```
(which is actually two separate optimizations.)

I haven't really tested the changes to the disabled passes though, just tried to keep them right.  Don't remember if they work or were just not worth the extra cycles...

-[Unknown]

### CODE REVIEW

File: Core/MIPS/IR/IRCompLoadStore.cpp [hunk 1]
```diff
@@ -46,113 +46,8 @@
...
-		default:
-			INVALIDOP;
-			return;
-		}
-
-		if (!load) {
-			// RAM(addrReg) = valueReg
-			ir.Write(IROp::Store32, valueReg, addrReg, ir.AddConstant(0));
-		}
+		// Should never get here, done in Comp_ITypeMem().
```
Comment (hrydgard) [inline Core/MIPS/IR/IRCompLoadStore.cpp]: Can't we just delete this whole function then?

**Comment by hrydgard** [comment]
Very neat.

For the interpreter, it might actually be faster to leave the LoadLeft etc ops behind when they can't be combined, but we definitely want to combine them when we later emit native code from the IR, to avoid duplicating that work for each backend.

**Comment by hrydgard** [comment]
Another thing. The current IR is nice for interpretation but I think if we really want to optimize whole functions with it eventually, we might want to raise it to tree-form, like most other IRs. That avoids all the silliness of having to move ops up and down in order to be able to do peephole optimizations, or various scans in passes - ops will always have their inputs accessible directly in the tree form. But that's for later of course, if at all.

File: Core/MIPS/IR/IRCompLoadStore.cpp (hunk 1 above)
Comment (unknownbrackets) [reply Core/MIPS/IR/IRCompLoadStore.cpp]: Ah, oops, I thought it was part of the jit interface or something.  Will remove in some hours.

-[Unknown]

**Review by hrydgard — APPROVED** [review]

**Comment by unknownbrackets** [comment]
Yeah, I'm a little worried about such optimizations since we (might) still have to keep filling regs (see: discard jr ra not working everywhere.)

Though, I wonder if we could use the analyst to detect if jal ever reads from those regs.  Would have to discard those results if we ever compiled code not analyzed, though.

-[Unknown]

**Pull request merged**